File: src/materials/dielectric.cpp
#include "dielectric.h"
#include "observable.h"
#include "rand.h"
#include <cmath>

Dielectric::Dielectric(float indexRefraction)
{
  this->tint = Color(1, 1, 1);
  this->indexRefraction = indexRefraction;
}

Dielectric::Dielectric(const Color& tint, float indexRefraction)
{
  this->tint = tint;
  this->indexRefraction = indexRefraction;
}

float
Dielectric::reflectance(float cosine, float refraction) const
{
  float r0 = (1 - refraction) / (1 + refraction);
  r0 *= r0;

  return r0 + (1 - r0) * pow((1 - cosine), 5);
}

std::optional<Pair>
Dielectric::scatter(const Ray& ray, const Observation& obs) const
{
  Color attenuation = tint;

  float refractionRatio =
    obs.frontFacing ? 1.0 / indexRefraction : indexRefraction;

  Vec unitDirection = ray.direction.normalize();

  float cosTheta = fmin(-unitDirection.dot(obs.normal), 1.0);
  float sinTheta = sqrt(1 - cosTheta * cosTheta);

  bool cannotRefract = refractionRatio * sinTheta > 1.0;
  Vec direction;
  if (cannotRefract || reflectance(cosTheta, refractionRatio) > randomFloat()) {
    direction = reflect(unitDirection, obs.normal);
  } else {
    direction = refract(unitDirection, obs.normal, refractionRatio);
  }

  Ray scattered = Ray(obs.point, direction);
  return std::make_pair(scattered, attenuation);
}

File: src/material/dielectric.rs
use super::*;
use std::sync::Arc;

#[derive(Debug, Clone, Copy)]
pub struct Dielectric {
    refraction_index_ratio: Float,
}

impl Dielectric {
    pub fn new(refraction_index: Float) -> Arc<Self> {
        Arc::new(Self {
            refraction_index_ratio: refraction_index,
        })
    }

    fn reflect(vector: Vector, normal: Vector) -> Vector {
        vector - 2.0 * vector.dot(&normal) * normal
    }

    fn refract(incoming: Vector, normal: Vector, ri_ratio: Float) -> Vector {
        let cos_theta = incoming.normalize().dot(&normal).min(1.0);
        let sin_theta = (1.0 - cos_theta * cos_theta).sqrt();

        // From some angles, refraction is impossible
        if ri_ratio * sin_theta > 1.0 {
            return Dielectric::reflect(incoming, normal);
        }

        let ray_perpendicular = ri_ratio * (incoming - cos_theta * normal);
        let ray_parallel = normal * (1.0 - ray_perpendicular.norm_squared()).abs().sqrt();

        ray_perpendicular - ray_parallel
    }
}

impl Material for Dielectric {
    fn scatter(&self, ray: &Ray, collision: &Collision) -> Option<Scatter> {
        let mut ratio = self.refraction_index_ratio;
        if collision.is_front_facing {
            ratio = 1.0 / ratio;
        }

        let unit_direction = ray.direction.normalize();
        let refracted = Dielectric::refract(unit_direction, collision.normal.normalize(), ratio);

        Some(Scatter {
            scattered: Ray::new(collision.point, refracted.normalize()),
            attenuation: WHITE,
        })
    }
}
